PR; Linq to RQL: ValueType .Parse() support

xinix00 opened the pull request:
To support valueType.Parse() for any conversion needed in your projection.

Based on Visitor Pattern, only removes the int/double/....Parse(), since conversion is already done in BLIT to Object. Pretty straight forward. Added a HashSet containing the ValueTypes, since Type.IsValueType isn't supported in dotnetstandard 1.3.

UnitTest included.

Commit:
PR; Linq to RQL: ValueType .Parse() support

Commit:
PR; Linq to RQL: ValueType .Parse() support, renamed UnitTest to better fit naming.

xinix00 committed:
Merge branch 'v4.0' into PR-003

ayende commented on test/FastTests/Client/QueriesWithCustomFunctions.cs:
```diff
@@ -1577,6 +1577,73 @@ select new
+                                };
+
+                    Assert.Equal("FROM Users as u SELECT { " +
+                        "IntParse : \"1234\", " +
```
But this outputs a _string_, not a value. 
That is probably not a good idea.
We want to send the raw value, not as a string.

ayende commented on test/FastTests/Client/QueriesWithCustomFunctions.cs:
```diff
@@ -1577,6 +1577,73 @@ select new
+                    Assert.Equal(3, queryResult.Count);
+
+
+                    Assert.Equal(1234, queryResult[0].IntParse);
```
This only works because json.net does type coercion, might not work for other clients.

xinix00 commented:
It's intend to work on things like: (where Id = purchase/100392123)
````
select new {
   EmployeeNumber = int.Parse(this.Id.Split("/")[1])
}
````

which would be:
````
select new {
    EmployeeNumber = this.Id.split("/")[1];
}
````
Which then would be converted to int, since the target was int, and JSON.NET will convert string to int (using int.Parse()). It looked a bit redundant to me to do:
````
select new {
    EmployeeNumber = parseInt(this.Id.split("/")[1]);
}
````

But I can make it as such:
````
                    Assert.Equal("FROM Users as u SELECT { " +
                        "IntParse : parseInt(\"1234\"), " +
                        "DoubleParse : parseFloat(\"1234\"), " +
                        "DecimalParse : parseFloat(\"1234\"), " +
                        "BoolParse : \"true\" == \"true\", " +
                        "CharParse : \"s\", " +
                        "ByteParse : parseInt(\"127\"), " +
                        "LongParse : parseInt(\"1234\"), " +
                        "SByteParse : parseInt(\"127\"), " +
                        "ShortParse : parseInt(\"1234\"), " +
                        "UintParse : parseInt(\"1234\"), " +
                        "UlongParse : parseInt(\"1234\"), " +
                        "UshortParse : parseInt(\"1234\") }", query.ToString());
````
So in a real life example you will get: (where PriceAsString = "$ 20.32")
````
FROM Users as u SELECT {
   PriceAsString: a.PriceAsString,
   PriceAsDecimal: parseFloat(a.PriceAsString.split(" ")[1])
}
````
Char is translated as such already in ConvertToJavascript (a char will be translated to the character in string). 

Though there are some differences between parseFloat and float.Parse():
````
JS: parseFloat("40 years") => 40
C#: float.Parse("40 years") => Bad format
````
Since JSON.NET is parsing it (or blittable, which is using convert. which will call int.Parse), when having ths:
````
select new {
     Test: int.Parse("2 years")
}
````
will be translated to 
````
{ 
    Test: "2 years"
}
````
Which will then get the exception you were expecting (Bad format) instead of 2. I left it out (int.Parse() etc.) since the current behavior is already converting string to int when needed. So I think it should be the original commit, or the one suggested above (or non at all ;))

Edit: That wasn't a cancel button, but a Close and Comment button...

xinix00 commented:
Maybe the confusion is about the unitTest, a better one would be:

xinix00 commented:
After some more thinking:
Test = int.Parse("123") + int.Parse("123")

would now be 123123, but when converted using parseInt, would be 246, so I create a commit with parseInt and parseFloat, since that would fit better.

Commit:
PR; Linq to RQL: ValueType Parse support, changes on review

ayende commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -430,22 +430,28 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
                 var method = methodCallExpression?.Method;
 
-                if (method == null || method.Name != "Parse" || !ValueTypes.Contains(method.DeclaringType))
+                if (method == null || method.Name != "Parse" || !ValueTypes.ContainsKey(method.DeclaringType))
```
`TryGetValue`, instead.
